### PR TITLE
Feature) New node<->libusb integration

### PR DESCRIPTION
**Pull request by jontore**
Responsible for communicating with camera, from node.
Using nodes NAPI to integrate a subset of libusb  with node.

### CODE REVIEW

File: cpp/Libusb.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+namespace libusb::internal {
+    struct Inner {
+        explicit Inner(libusb_context * ctx) : ctx(ctx) {}
+        Inner(Inner &&) = delete;
+        Inner(Inner const &) = delete;
+        Inner& operator=(Inner const &) = delete;
+        Inner& operator=(Inner &&) = delete;
+        ~Inner() {
+            std::cout << "Libusb: libusb_exit" << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: delete

File: cpp/Libusb.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    struct Open_inner {
+        Open_inner(std::shared_ptr<Inner> ctx, libusb_device_handle * devh, bool silent)
+            : ctx(std::move(ctx)), devh(devh), silent(silent) {}
+        Open_inner(Open_inner &&) = delete;
+        Open_inner(Open_inner const &) = delete;
+        Open_inner operator=(Open_inner const &) = delete;
+        Open_inner operator=(Open_inner &&) = delete;
+        ~Open_inner() {
+            if (!silent) {
+                std::cout << "Libusb: libusb_close" << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: delete

File: cpp/Libusb.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        bool const silent;
+    };
+};
+
+std::variant<Libusb, libusb::Error> Libusb::init()
+{
+    libusb_context *ctx;
+    int const r = libusb_init(&ctx);
+    if (r != 0) {
+        std::cerr << "Libusb: libusb_init failed: " << r << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: print string instead of number

File: cpp/Libusb.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    return Libusb(inner->ctx);
+}
+
+libusb::Open_device::Open_device(std::shared_ptr<libusb::internal::Inner> ctx, libusb_device_handle * devh, bool silent)
+    : inner(std::make_shared<libusb::internal::Open_inner>(ctx, devh, silent)) {}
+
+std::variant<libusb::Claimed_interface, libusb::Error> libusb::Open_device::claim_interface(int interface_number) {
+    int const r = libusb_claim_interface(inner->devh, interface_number);
+    if (r < 0) {
+        std::cerr << "Libusb: libusb_claim_interface failed: " << r << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: string instead of number

File: cpp/Libusb.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        libusb::Error err(r);
+        switch (r) {
+        case LIBUSB_ERROR_TIMEOUT:
+            if (actual_length != 0 && actual_length != length && (ep_in & 0x80) != 0) {
+                //std::cerr << "Uh, oh. Transferred " << actual_length << " bytes on timeout" << std::endl;
+                return actual_length;
+            }
+            break;
+        default:
+            std::cerr << "Libusb: libusb_bulk_transfer failed: " << err.get_message() << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: don't print here.

File: cpp/Libusb.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    ret.resize(r);
+    return ret;
+}
+
+libusb::Claimed_interface::~Claimed_interface() {
+    if (inner == nullptr) {
+        return;
+    }
+    std::cout << "Libusb: libusb_release_interface" << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: delete

File: cpp/Libusb.cpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    int const r = libusb_release_interface(inner->devh, interface_number);
+    if (r != 0) {
+        std::cerr << "Libusb: libusb_release_interface failed: " << r << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.cpp]: Don't print here. Ignore the error.

File: scripts/build_linux.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash -l
```
Comment (jontore) [inline scripts/build_linux.sh]: We should rebuild all the binaries.

File: cpp/Libusb.hpp [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+			//std::cout << "Libusb: libusb_unref_device " << dev << std::endl;
+			libusb_unref_device(dev);
+		}
+
+        std::variant<Open_device, Error> open(bool silent=false) {
+            assert(dev != nullptr);
+			assert(inner != nullptr);
+            libusb_device_handle *devh = nullptr;
+            if (!silent) {
+                std::cout << "Libusb: libusb_open" << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.hpp]: don't print here

File: cpp/Libusb.hpp [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+            assert(r == 0); // Always succeeds, according to docs.
+            return ret;
+        }
+
+        std::variant<Config_descriptor, Error> get_active_config_descriptor() const {
+			assert(inner != nullptr);
+            libusb_config_descriptor *config = nullptr;
+            int const r = libusb_get_active_config_descriptor(dev, &config);
+            if (r != 0) {
+                std::cerr << "Libusb: libusb_get_active_config_descriptor failed: " << r << std::endl;
```
Comment (kristofferkoch) [inline cpp/Libusb.hpp]: print string, not error number

File: package.json [hunk 11]
```diff
@@ -34,13 +34,15 @@
     "url": "git+https://github.com/Huddly/device-api-usb.git"
   },
   "scripts": {
-    "build": "npm run build-ts && npm run tslint && npm run test",
+    "build": "npm run build-ts && npm run tslint && npm run napi",
     "build-ts": "tsc --declaration",
     "publish-package": "npm run build && npm publish",
```
Comment (jontore) [inline package.json]: This needs to move the prebuilds into the lib folder for publishing.

File: cpp/bulk_usb.cpp [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    if (cookie_num == 0) {
+        Napi::TypeError::New(env, "Cookie can not be 0").ThrowAsJavaScriptException();
+        return env.Undefined();
+    }
+    Usb_cookie const cookie(cookie_num);
+    auto js_cb = info[1].As<Napi::Function>();
+    auto cbinfo = std::make_shared<CallbackInfo>(env, js_cb);
+    async.ref();
+    worker_arg->open_device(cookie, [env, cbinfo=std::move(cbinfo)](int error, Usb_cookie handle){
+        std::cout << "in openDevice" << std::endl;
```
Comment (kristofferkoch) [inline cpp/bulk_usb.cpp]: don't print here

File: cpp/bulk_usb.cpp [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+        Napi::TypeError::New(env, "Cookie can not be 0").ThrowAsJavaScriptException();
+        return env.Undefined();
+    }
+    Usb_cookie const cookie(cookie_num);
+    auto js_cb = info[1].As<Napi::Function>();
+    auto cbinfo = std::make_shared<CallbackInfo>(env, js_cb);
+    //std::cout << "closeDevice cookie " << cookie << std::endl;
+    async.ref();
+    worker_arg->close_device(cookie, [env, cbinfo=std::move(cbinfo)](int error) {
+        std::cout << "in closeDevice" << std::endl;
```
Comment (kristofferkoch) [inline cpp/bulk_usb.cpp]: remove print

File: src/bulkusbdevice.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+          newDevices.forEach(newDevice => this._onAttaches.forEach(cb => cb(newDevice)));
+          return resolve(Object.freeze(ret));
+        });
+      });
+  }
+
+  private async _pollLoop() {
+    this._isPolling = true;
+    console.log('Starting BulkUsb poll loop');
```
Comment (kristofferkoch) [inline src/bulkusbdevice.ts]: remove print

File: src/bulkusbdevice.ts [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    for (;;) {
+      // Guarantee that the actual _listDevices call is done after the listDevices call.
+      // Not before (cached), and not while listing, but after.
+      const toResolve = this._pollingListResolve;
+      this._pollingListResolve = [];
+      try {
+        const devices = await this._listDevices();
+        toResolve.map(cb => cb(devices));
+      } catch (e) {
+        console.log('BulkUsb attach poll loop got error:', e);
```
Comment (kristofferkoch) [inline src/bulkusbdevice.ts]: Should use `this.logger`?

File: src/bulkusbdevice.ts [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+  listDevices(): Promise<ReadonlyArray<BulkUsbDevice>> {
+    if (!this._isPolling) {
+      return this._listDevices();
+    }
+    return new Promise((resolve) => {
+      this._pollingListResolve.push(resolve);
+    });
+  }
+
+  async onAttach(cb: (dev: BulkUsbDevice) => Promise<void>) {
```
Comment (kristofferkoch) [inline src/bulkusbdevice.ts]: missing return type `Promise<void>`?

File: src/manager.ts [hunk 17]
```diff
@@ -118,44 +46,47 @@ export default class DeviceDiscoveryManager implements IDeviceDiscovery {
...
+      return;
+    }
+    this.attachedDevices = this.attachedDevices.filter(d => !removedDevice.equals(d));
+    this.eventEmitter.emit('DETACH', removedDevice.serialNumber);
+  }
 
-    this.updateCache(newDevices, removedDevices);
-    return { devices: usbDevices, newDevices: newDevices, removedDevices: removedDevices };
+  async deviceList(): Promise<{devices: BulkUsbDevice[]}> {
+    // Fixme: Do dummy attach to init polling
```
Comment (jontore) [inline src/manager.ts]: What should we do here?

File: src/errstr.ts [hunk 18]
```diff
@@ -0,0 +1,10 @@
+export function errstr(n: any): Error {
+  switch (n) {
+    case -1: return new Error('LIBUSB_ERROR_IO');
```
Comment (kristofferkoch) [inline src/errstr.ts]: Add all possible libusb errors.

File: src/transport.ts [hunk 19]
```diff
@@ -74,110 +92,119 @@ export default class NodeUsbTransport extends EventEmitter implements ITransport
...
+        reject(e);
+      }
+    }
+  }
+
+  async readMessage(): Promise<void> {
+    let headerBuffer: Buffer;
+    do {
+      try {
+        headerBuffer = await this.endpoint.read(4096, HEADER_TIMEOUT_MS);
```
Comment (jontore) [inline src/transport.ts]: magic number?

File: src/transport.ts [hunk 20]
```diff
@@ -74,110 +92,119 @@ export default class NodeUsbTransport extends EventEmitter implements ITransport
...
-    this.inEndpoint.on('error', (err) => {
-      this.logger.warn(`Received an error message on read loop! ${err}`);
-      this.close();
-    });
 
-    this.inEndpoint.on('end', () => {
-      // Polling has stopped!
-    });
+  async startListen(): Promise<void> {
+    throw new Error('----------- SHOLUD NOT HAPPEND legacy listen --------------');
```
Comment (jontore) [inline src/transport.ts]: Change error message to deprecated error.

**Pull request merged**